src/key.cpp

```diff
@@ -354,8 +375,21 @@ bool CKey::SetCompactSignature(uint256 hash, const std::vector<unsigned char>& v
     if (nV<27 || nV>=35)
         return false;
     ECDSA_SIG *sig = ECDSA_SIG_new();
-    BN_bin2bn(&vchSig[1],32,sig->r);
-    BN_bin2bn(&vchSig[33],32,sig->s);
+    if (!sig) return false;
+
+    BIGNUM *sig_r = BN_new(), *sig_s = BN_new();
+
+    BN_bin2bn(&vchSig[1],32,sig_r);
+    BN_bin2bn(&vchSig[33],32,sig_s);
+
+    #if OPENSSL_VERSION_NUMBER > 0x1000ffffL
+    ECDSA_SIG_set0(sig, sig_r, sig_s);
+    #else
+    if (sig->r) BN_clear_free(sig->r);
+    if (sig->s) BN_clear_free(sig->s);
+    sig->r = sig_r;
+    sig->s = sig_s;
```

Comment:
This changes the behavior a bit. I don't know the internals of openssl, but if it ever keeps another reference to `r` and `s`  then this code breaks it. It's most likely not the case, but keeping the old code as it was would be simpler and less risky.